### PR TITLE
[exporter/syslog] send syslog messages in batches

### PR DESCRIPTION
**Pull request by andrzej-stencel**
**Description:**

This changes the behavior of the Syslog exporter to send each batch of Syslog messages in a single request (with messages separated by newlines), instead of sending each message in a separate request and closing the connection after each message.

The batching only happens when using TCP. For UDP, each syslog message is still sent in a separate request, as defined by [the spec](https://datatracker.ietf.org/doc/html/rfc5426#section-3.1).

This also significantly refactors (and hopefully simplifies) the exporter's code, extracting the code that formats the syslog messages from the `sender` type into separate `formatter` types. Hopefully this will make the development of this component easier.

**Link to tracking Issue:**

- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21244

**Testing:**

The unit tests have been updated to reflect the refactored codebase. The integration tests introduced in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27464 are unchanged, as the format of the output messages hasn't changed.

**Documentation:**

No documentation updates.

### CODE REVIEW

**Review by songy23 — APPROVED** [review]

File: exporter/syslogexporter/exporter.go
Comment (rnishtala-sumo) [inline exporter/syslogexporter/exporter.go:80]: should we have a configurable limit on the payload size, now that we're batching log messages?

Comment (astencel-sumo) [reply exporter/syslogexporter/exporter.go:80]: Good point that it should be possible to limit the payload size. Users should be able to use the collector's built-in batching for this. Currently this functionality is provided the [Batch processor](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/batchprocessor). In future, batching hopefully will be a part of the [exporter helper](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper).

**Review by rnishtala-sumo — APPROVED** [review]

File: exporter/syslogexporter/exporter.go
Comment (rnishtala-sumo) [reply exporter/syslogexporter/exporter.go:80]: oh right, I forgot that this is an exporter and not the receiver.

**Comment by astencel-sumo** [comment]
I have changed the implementation to only batch messages when using TCP. As correctly pointed out by @kkujawa-sumo (thank you!), when sending over UDP, each message must be sent in a separate request: https://datatracker.ietf.org/doc/html/rfc5426#section-3.1

> Each syslog UDP datagram MUST contain only one syslog message

File: exporter/syslogexporter/formatter.go
Comment (rnishtala-sumo) [inline exporter/syslogexporter/formatter.go:18]: should the formatter interface have a validation function that returns an error in the following scenarios

- a required field is empty
- The value for a field is not in the right format. For example:
   ```
   The PRI part MUST have three, four, or five characters and will be
   bound with angle brackets as the first and last characters.  The PRI
   part starts with a leading "<" ('less-than' character, %d60),
   followed by a number, which is followed by a ">"
```

Comment (kkujawa-sumo) [reply exporter/syslogexporter/formatter.go:18]: > should the formatter interface have a validation function that returns an error in the following scenarios
> 
> * a required field is empty
> * The value for a field is not in the right format. For example:
>   ```
>   The PRI part MUST have three, four, or five characters and will be
>   bound with angle brackets as the first and last characters.  The PRI
>   part starts with a leading "<" ('less-than' character, %d60),
>   followed by a number, which is followed by a ">"
>   ```

`format` function will always add the PRI part e.g. [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27799/files#diff-fd44958375c0625e37038008e6afcd03ad386b6e60db4fb1d4d1fdccb1c4a27dR30)

for example [time.Format ](https://pkg.go.dev/time#Time.Format) doesn't return any error

**Review by kasia-kujawa — APPROVED** [review]

File: exporter/syslogexporter/formatter.go
Comment (rnishtala-sumo) [reply exporter/syslogexporter/formatter.go:18]: makes sense, the formatter ensures a valid format for PRI and other fields, I wonder what kind of validation logic exists in the syslog receiver. The exporter probably doesn't need to do any field validation (apart from populating default values).

File: .chloggen/syslog-exporter-single-request.yaml
Comment (astencel-sumo) [reply .chloggen/syslog-exporter-single-request.yaml:10]: Good point, updated.

File: .chloggen/syslog-exporter-single-request.yaml
Comment (astencel-sumo) [reply .chloggen/syslog-exporter-single-request.yaml:18]: Updated.

File: exporter/syslogexporter/formatter.go
Comment (astencel-sumo) [reply exporter/syslogexporter/formatter.go:18]: I initially implemented the `formatter`'s `format` function to return `(string, error)` but then it occurred that the implementation doesn't ever return an error. As there is no erroneous input, there's no need for a validation function.

File: exporter/syslogexporter/rfc5424_formatter_test.go
Comment (astencel-sumo) [reply exporter/syslogexporter/rfc5424_formatter_test.go:19]: The exporter never fails on formatting the message. It will use default values if anything is missing.

**Comment by astencel-sumo** [comment]
@songy23 anything missing from marking it as `ready to merge`? Not sure if this is the way to signal to maintainers that this can be merged.

**Comment by songy23** [comment]
Added the label, cc @open-telemetry/collector-contrib-maintainer please merge 🙏

**Pull request merged**
